### PR TITLE
build: avoid duplicated build of hwloc and mpl

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Avoid duplicated build of hwloc and mpl in hydra by reusing the libtool package build in the main mpich.

Also, add `-quick` option to autogen to skip many unneeded submodule initializations.

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
Before this PR:
```
./autogen.sh && ./configure --with-device=ch4:ofi --disable-shared && make install
```
On my local work station: `11m12sec`

After this PR
```
./autogen -quick -with-ofi -with-yaksa -with-hydra -with-json -with-romio
./configure --with-device=ch4:ofi --disable-shared
make install
```
On my local work station: `9m23sec`

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/warnings/auto

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

File: autogen.sh [hunk 1]
```diff
@@ -91,6 +91,8 @@ echo "done"
 ## Initialize variables to default values (possibly from the environment)
 ########################################################################
 
+do_quick=no
+
```
Comment (raffenet) [inline autogen.sh]: i think this leaked from another commit

File: autogen.sh [hunk 2]
```diff
@@ -131,7 +131,9 @@ for arg in "$@" ; do
 	-echo)
 	    set -x
 	    ;;
-	
+        -release)
```
Comment (raffenet) [inline autogen.sh:155]: is this used?

File: autogen.sh (hunk 1 above)
Comment (hzhou) [reply autogen.sh]: cleaned up.

File: autogen.sh (hunk 2 above)
Comment (hzhou) [reply autogen.sh]: removed.

**Comment by hzhou** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by wesbland** [comment]
> Before this PR:
> 
> ```
> ./autogen.sh && ./configure --with-device=ch4:ofi --disable-shared && make install
> ```
> 
> On my local work station: `11m12sec`
> 
> After this PR
> 
> ```
> ./autogen -quick -with-ofi -with-yaksa -with-hydra -with-json -with-romio
> ./configure --with-device=ch4:ofi --disable-shared
> make install
> ```
> 
> On my local work station: `9m23sec`

So why not make `-quick` the default? Also, did you add the requirement for `-with-ofi` and co.?

**Comment by hzhou** [comment]
> 
> 
> > Before this PR:
> > ```
> > ./autogen.sh && ./configure --with-device=ch4:ofi --disable-shared && make install
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On my local work station: `11m12sec`
> > After this PR
> > ```
> > ./autogen -quick -with-ofi -with-yaksa -with-hydra -with-json -with-romio
> > ./configure --with-device=ch4:ofi --disable-shared
> > make install
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On my local work station: `9m23sec`
> 
> So why not make `-quick` the default? Also, did you add the requirement for `-with-ofi` and co.?

`-quick` turns most of the optional off and you have to know what you are doing to turn individual work back on.

**Comment by wesbland** [comment]
> `-quick` turns most of the optional off and you have to know what you are doing to turn individual work back on.

That's fine, but if it's really that much better and there's no reason not to have it on, we can change that. We're all powerful programmers here. 😄

**Comment by hzhou** [comment]
> 
> 
> > `-quick` turns most of the optional off and you have to know what you are doing to turn individual work back on.
> 
> That's fine, but if it's really that much better and there's no reason not to have it on, we can change that. We're all powerful programmers here. 😄

The reason is how do you expect a random user or developer know what they are doing?

**Comment by wesbland** [comment]
That's exactly my point.

Maybe I can better state my question. What is the downside to not making `-quick` the default? Is there some case that it breaks things?

**Comment by hzhou** [comment]
> 
> 
> That's exactly my point.
> 
> Maybe I can better state my question. What is the downside to not making `-quick` the default? Is there some case that it breaks things?

If we make `-quick` the default, a random user or developer is likely to encounter situations that "why" xxx doesn't work, deep down the path.

Not making `-quick` the default, `autogen` will do much extra work, but it should always work.

You restated question really confuses me, which side are you on? 😃

**Comment by wesbland** [comment]
> You restated question really confuses me, which side are you on? 😃

I'm on the side of making `-quick` default if possible. I want things to be faster by default unless it causes problems.

> If we make `-quick` the default, a random user or developer is likely to encounter situations that "why" xxx doesn't work, deep down the path.

Let's ignore the "user" case. We have always made the assumption that users are building from tarballs, which means they don't run `autogen.sh`. Let's try to make it easier for "developers" (i.e., me 😁).

In what situation will `-quick` break?

**Comment by hzhou** [comment]
> > If we make `-quick` the default, a random user or developer is likely to encounter situations that "why" xxx doesn't work, deep down the path.
> 
> Let's ignore the "user" case. We have always made the assumption that users are building from tarballs, which means they don't run `autogen.sh`. Let's try to make it easier for "developers" (i.e., me 😁).
> 

There are more people trying to build from git repository than we are aware. We can assume more expectations from them, but we cannot ignore them.

> In what situation will `-quick` break?

If you just `autogen -quick`, all your build probably will fail. For example, hydra and libfabric is not autoconf'd.

I was mainly thinking about CI jobs, which we can hard code for each job. For manual checkout, I am still expecting everyone just do the default "slow" `autogen`. ... Oh, I see. The `-quick` options only affects the autogen. Both the `configure` and `make` will still avoid double build `mpl` and `hwloc`.

To have `autogen` smartly choose components, it needs to know your configure options.

**Review by raffenet — APPROVED** [review]
FWIW, I think keeping the quick option as opt-in is good for now.

**Pull request merged**